### PR TITLE
Extract reusable location header component

### PR DESCRIPTION
**Pull request by bryangaledxw**
# Changes in this PR

This PR first does some tidy up, most significantly adding better type safety to Page.verifyOnPage, which then necessitates some fixes. We then add then use a reusable location header component, replacing a lot of repeated code. This PR should cause no user-visible changes

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

**Comment by bryangaledxw** [comment]
[Extract reusable "location header" component](https://trello.com/c/plGvRu9W/853-extract-reusable-location-header-component)

File: cypress_shared/components/locationHeader.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    cy.get('.location-header').within(() => {
+      if (room) {
+        cy.get('p').should('contain', room.name)
+      }
+
+      if (premises) {
+        cy.get('p').should('contain', premises.name)
+
+        if (!this.hideAddress) {
```
Comment (patrickjfl) [inline cypress_shared/components/locationHeader.ts]: Is it worth also checking the else condition here?

**Review by patrickjfl — APPROVED** [review]
LGTM, although locally for me there is no location header on the 'Cancel booking' or 'Mark booking as active' screens - is this a known issue? Seems like it would be as necessary/helpful as anywhere else...

**Comment by bryangaledxw** [comment]
@patrickjfl Yep, so that's the next PR. This PR has those pages using the shared component, but only passing in the CRN. The next PR will actually pass in the premises and room. There's extra that needs to come from the API on those pages, so I thought that worth doing in its own PR

File: cypress_shared/components/locationHeader.ts (hunk 1 above)
Comment (bryangaledxw) [reply cypress_shared/components/locationHeader.ts]: Yep, good call, will add

**Pull request merged**